src/mlpack/tests/lmnn_test.cpp

```diff
@@ -0,0 +1,551 @@
+  // Get gradients for the current parameters.
+  arma::mat orgGradient, gradient, estGradient;
+  function.Gradient(coordinates, orgGradient);
+
+  estGradient = arma::zeros(orgGradient.n_rows, orgGradient.n_cols);
+
+  // Compute numeric approximations to gradient.
+  for (size_t i = 0; i < orgGradient.n_elem; ++i)
+  {
+    double tmp = coordinates(i);
+
+    // Perturb parameter with a positive constant and get costs.
+    coordinates(i) += eps;
+    double costPlus = function.Evaluate(coordinates);
+
+    // Perturb parameter with a negative constant and get costs.
+    coordinates(i) -= (2 * eps);
+    double costMinus = function.Evaluate(coordinates);
+
+    // Restore the parameter value.
+    coordinates(i) = tmp;
+
+    // Compute numerical gradients using the costs calculated above.
+    estGradient(i) = (costPlus - costMinus) / (2 * eps);
+  }
+
+  // Estimate error of gradient.
+  return arma::norm(orgGradient - estGradient) /
+      arma::norm(orgGradient + estGradient);
+}
+
+BOOST_AUTO_TEST_CASE(LMNNFunctionGradientTest)
+{
+  // Useful but simple dataset with six points and two classes.
+  arma::mat dataset        = "-0.1 -0.1 -0.1  0.1  0.1  0.1;"
+                             " 1.0  0.0 -1.0  1.0  0.0 -1.0 ";
+  arma::Row<size_t> labels = " 0    0    0    1    1    1   ";
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6, 1);
+
+  // 10 trials with random positions.
+  for (size_t i = 0; i < 10; ++i)
+  {
+    arma::mat coordinates(2, 2, arma::fill::randn);
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(LMNNFunctionGradientTest2)
+{
+  // Useful but simple dataset with six points and two classes.
+  arma::mat dataset        = "-0.1 -0.1 -0.1  0.1  0.1  0.1;"
+                             " 1.0  0.0 -1.0  1.0  0.0 -1.0 ";
+  arma::Row<size_t> labels = " 0    0    0    1    1    1   ";
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6, 1);
+
+  // 10 trials with random positions.
+  for (size_t i = 0; i < 10; ++i)
+  {
+    arma::mat coordinates(2, 2, arma::fill::randu);
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(LMNNFunctionGradientTest3)
+{
+  arma::mat dataset;
+  arma::Row<size_t> labels;
+  data::Load("iris.csv", dataset);
+  data::Load("iris_labels.txt", labels);
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6, 1);
+
+  // 10 trials with random positions.
+  for (size_t i = 0; i < 10; ++i)
+  {
+    arma::mat coordinates(dataset.n_rows, dataset.n_rows, arma::fill::randn);
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(LMNNFunctionGradientTest4)
+{
+  arma::mat dataset;
+  arma::Row<size_t> labels;
+  data::Load("iris.csv", dataset);
+  data::Load("iris_labels.txt", labels);
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6, 1);
+
+  // 10 trials with random positions.
+  for (size_t i = 0; i < 10; ++i)
+  {
+    arma::mat coordinates(dataset.n_rows, dataset.n_rows, arma::fill::randu);
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
```

Comment:
I think it would be a good idea to check that we can run LMNN with a low-rank distance matrix and we still get reasonable results.  You could create a high-dimensional random dataset (maybe two Gaussians or something, one for each class) and then run LMNN twice, once with a low-rank matrix and once with a normal matrix, and then make sure the KNN accuracies are comparable at the end of the two runs.  What do you think?

You could do that before merge or we can open a ticket for it, I am not picky.

Reply:
sure, I will do that.

Reply:
Done!